enumerables.rb

```diff
@@ -1,41 +1,151 @@
-    def my_each
-      return to_enum(:my_each) unless block_given?
-      
-      i = 0
-      array = to_a
-      while i < array.length
-        yield(array[i])
-        i+=1
+  def my_each
+    new_array = [] unless block_given?
+    i = 0
+    a = Array self
+    while i < a.size
+      yield(a[i]) if block_given?
+      new_array << a[i] unless block_given?
+      i += 1
     end
-    array
+    return Enumerator.new(new_array) unless block_given?
+
+    self
   end
- 
 
-def my_each_with_index
-    return to_enum(:my_each_with_index) unless block_given?
-    
+  def my_each_with_index
+    new_array = [] unless block_given?
     i = 0
-    array = to_a
-    while i < array.length
-      yield(array[i], i)
-      i+=1
+    a = Array self
+    while i < size
+      yield(a[i], i) if block_given?
+      new_array << a[i] unless block_given?
+      i += 1
+    end
+    return Enumerator.new(new_array) unless block_given?
+
+    self
   end
-  array
-end
 
-def my_select
-    return to_enum(:my_select) unless block_given?
+  def my_select
+    return my_each unless block_given?
+
     new_array = []
-  my_each do |x|
-new_array.push(x) if yield(x) == true
+    my_each { |n| new_array.push(n) if yield(n) }
+    new_array
+  end
+
+  def my_all?(*args)
+    result = true
+    if block_given?
+      my_each do |i|
+        result = false unless yield(i)
+      end
+    elsif !args[0].nil?
+      my_each do |i|
+        result = false unless args[0] === i
+      end
+    elsif empty? && args[0].nil?
+      result = true
+    else
+      my_each do |i|
+        result = false unless i
+      end
+    end
+    result
+  end
+
+  def my_any?(*args)
+    result = false
+    if block_given?
+      my_each do |i|
+        result = true if yield(i)
+      end
+    elsif !args[0].nil?
+      my_each do |i|
+        result = true if args[0] === i
+      end
+    elsif empty? && args[0].nil?
+      result = false
+    else
+      my_each do |i|
+        result = true if i
+      end
+    end
+    result
+  end
+
+  def my_none?(args = nil)
+    result = if block_given?
+               !my_any? { |i| yield i }
+             else
+               !my_any?(args)
+             end
+    result
+  end
+
+  def my_count(item = false)
+    count = 0
+
+    if !block_given?
+      return size unless item
+
+      my_each { |n| count += 1 if n == item }
+    elsif block_given? && !item
+      my_each { |n| count += 1 if yield(n) }
+    elsif block_given? && item
+      my_each { |n| count += 1 if n == item }
+    end
+    count
   end
-  new_array
-end
-end
-  
 
-puts [1,2,3,4,5,6].my_select(&:even?)
+  def my_map(args_proc = nil)
+    return to_enum :my_map unless block_given? || args_proc.class == Proc
+    new_arr = []
+    new_object = to_a
+    if !args_proc.nil?
+      new_object.my_each do |i|
+        new_arr.push(proc.call(i))
+      end
+    else
+      new_object.my_each do |i|
+        new_arr.push(yield(i))
+      end
+    end
+    new_arr
+  end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so let's work on it as it was requested by previous reviewers.

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```